core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -88,12 +88,12 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
 
       new IO.SyncEffect(() => g(io.effect()))
 
-    case IO.Tags.Fail =>
-      val io = self.asInstanceOf[IO.Fail[E]]
+    case IO.Tags.Terminate =>
+      val io = self.asInstanceOf[IO.Terminate[E]]
 
-      new IO.Fail(f(io.error), io.defects)
+      new IO.Terminate(io.cause.map(f))
 
-    case _ => new IO.Redeem(self, (e: E) => new IO.Fail(f(e), Nil), (a: A) => new IO.Strict(g(a)))
+    case _ => new IO.Redeem(self, (cause: Cause[E]) => new IO.Terminate(cause.map(f)), (a: A) => new IO.Strict(g(a)))
```

Comment:
This logic for short-circuiting is going to become more tricky. I'd replace all of this with:

```scala
bimap(f, g) = self.redeemPure(f, g)
```

Reply:
`redeemPure` needs `f` and `g` to produce the same type, but here in `bimap` they produce different types.